File: RSUtils/NSObject+dispatch.m
//
//  NSObject+dispatch.m
//  RSUtils
//
//  Created by mhtran on 9/25/17.
//  Copyright © 2017 mhtran. All rights reserved.
//

#import "NSObject+dispatch.h"

void execOnMain( void (^block)() ){
    dispatch_async( dispatch_get_main_queue(), block );
}

void delayToMain( double seconds, void (^block)() ){
    dispatch_after( dispatch_time( DISPATCH_TIME_NOW, ( int64_t )( seconds * NSEC_PER_SEC )), dispatch_get_main_queue(), block );
}

@implementation NSObject (dispatch)

+( void )execOnMain:( void (^)() )block; {
    execOnMain( block );
}

-( void )execOnMain:( void (^)() )block; {
    execOnMain( block );
}

+( void )delayToMain:( double )seconds exec:( void (^)() )block {
    delayToMain( seconds, block );
}

-( void )delayToMain:( double )seconds exec:( void (^)() )block {
    delayToMain( seconds, block );
}

@end

File: RSUtils/NSObject+dispatch.swift
//
//  NSObject+dispatch.swift
//  RSUtils
//
//  Created by mhtran on 11/2/17.
//  Copyright © 2017 mhtran. All rights reserved.
//
import Foundation
extension NSObject {
    class func exec(onMain block: @escaping () -> Void) {
        execOnMain(block: block())
    }
    class func delay(toMain seconds: Double, exec block: @escaping () -> Void) {
        delayToMain(seconds: seconds, block: block())
    }
    func exec(onMain block: @escaping () -> Void) {
        execOnMain(block: block())
    }
    func delay(toMain seconds: Double, exec block: @escaping () -> Void) {
        delayToMain(seconds: seconds, block: block())
    }
}
func execOnMain(block: Void) {
    DispatchQueue.main.async {block}
}
func delayToMain(seconds: Double, block: Void) {
    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(ino64_t(seconds * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC)) {
        block
    }
}
